Add Dockerfile

JanKoppe opened the pull request:
This is a work in progress.

Adds a Dockerfile based on Alpine Linux 3.4.

JanKoppe commented:
This seems to be working just fine. Can be merged imho!

strugee commented:
@evanp do you want to do a review on this? You're more familiar with Docker than me

strugee commented on docker/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+
+The Dockerfile also allows to define the UID and GUID of the `pumpio` user which runs pump.io via build arguments.
+```
````
Newline before this one

strugee commented on docker/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Running
+
+By default this Dockerfile only includes the `databank-mongodb` driver. Therefore you will have to add the configuration parameters for the mongodb. The recommended way to configure pump.io is to configure the options via environment variables, as described in the projects README. You can see some example variables in the `docker-compose.yml` file. This file wil also show a basic working version  ...
```
s/the mongodb/MongoDB/

strugee replied:
s/projects/project's/

strugee replied:
Typo: `wil` should be `will`

strugee commented on docker/README.md:
````diff
@@ -0,0 +1,39 @@
+By default this Dockerfile only includes the `databank-mongodb` driver. Therefore you will have to add the configuration parameters for the mongodb. The recommended way to configure pump.io is to configure the options via environment variables, as described in the projects README. You can see some example variables in the `docker-compose.yml` file. This file wil also show a basic working version  ...
+
+Running via `docker-swarm`:
+```
````
Newline before this one

strugee commented on docker/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+
+Stopping and removing:
+```
````
Ditto

strugee commented on docker/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+docker-compose down -v
+```
+Notice the `-v` parameter, as the docker-compose file defines volumes. Without this parameter they will be left behind.
````
Newline before this one

strugee replied:
Also should "docker-compose" be surrounded by `s? I dunno.

strugee commented on docker/README.md:
```diff
@@ -0,0 +1,39 @@
+Notice the `-v` parameter, as the docker-compose file defines volumes. Without this parameter they will be left behind.
+
+
+If you have Docker Swarm active, you can also deploy this aas a Service Stack
```
`as` instead of `aas` and this should probably have a `:` at the end

strugee commented on docker/README.md:
````diff
@@ -0,0 +1,39 @@
+
+
+If you have Docker Swarm active, you can also deploy this aas a Service Stack
+```
````
Newline

strugee commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+      - NODE_ENVIRONMENT=production
+      - PUMPIO_ADDRESS=0.0.0.0
+      - PUMPIO_DATADIR=/var/local/pumpio
+      - PUMPIO_ENABLEUPLOADS=true
```
`PUMPIO_ENABLE_UPLOADS` per https://github.com/pump-io/pump.io#environment

strugee commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+networks:
+  pumpio_net:
+
+
```
Take out this trailing newline

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+RUN apk add --no-cache graphicsmagick openssl nodejs python make g++ git \
+     && mkdir -p "${PUMP_LOCATION}" \
+     && git clone --recursive --depth 1 --branch "${PUMP_GIT_TAG}" "${PUMP_GIT_REPO}" "${PUMP_LOCATION}" \
```
As I mentioned in #1307 it seems better/more intuitive to me to build the Docker image from the local source instead of cloning. Is this not how it's usually done? You seem to know more than me soooo...

strugee replied:
Also I'm assuming this is just shell? Pretty sure this would better be `$PUMPIO_GIT_TAG` (without the `{}`), etc.

aurbot-user replied:
Yes, this is shell. In this case the curly braces can be left out, but it is common good practice to include them to avoid accidental misinterpretation of the variable name. (e.g. (`$PUMP_GIT_TAGsomeotherthing`)

Yepp, building from local source is a good idea. I tend to use that git clone style because then the whole Dockerfile itself is neatly portable. :)

JanKoppe replied:
Yes, this is just shell. In this case the curly braces make no difference, but it's common good practice to include them, to avoid misinterpretation of the variable name.

I usually like doing the git clone style because it makes the Dockerfile more portable, but sure, you can just build this out of the repository itself with a `COPY` statement. Not sure if you would have to move the Dockerfile to the root of the repository thought.

Also, I do remember now why I did that in the first place here - I started the Dockerfile for my own use before it was suggested that I make a PR in IRC. :)

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+     && adduser -S -D -H -G "pumpio" -h "/usr/local/pumpio" -u "${PUMPIO__UID}" "pumpio" \
+     && chown -R "pumpio:pumpio" /usr/local/pumpio \
+     && mkdir -p /usr/local/bin \
+     && ln -s ${PUMP_LOCATION}/bin/pump /usr/local/bin/pumpio \
```
Probably this should link to `pump` instead of `pumpio` a) for consistency with the npm package and b) since you're calling `pump` down below

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+     && npm install databank-mongodb \
+     && addgroup -S -g "${PUMPIO__GUID}" "pumpio" \
+     && adduser -S -D -H -G "pumpio" -h "/usr/local/pumpio" -u "${PUMPIO__UID}" "pumpio" \
+     && chown -R "pumpio:pumpio" /usr/local/pumpio \
```
Why is this in `/usr/local`? `/var/local` would probably be better

JanKoppe replied:
This is just so if you would ever want to drop into a shell inside the container your home would be the place that pump.io is located at. Just for convenience, though it does clash a bit with LFS.

JanKoppe replied:
wait, scratch that comment, thought you were referring to L21.

strugee replied:
No, I was. Though I think I got it confused - `/usr/local/pumpio` is supposed to be the place where the source is, right? Not where `datadir` is? In that case it should be left as `root:root`. No reason to make it owned by the pump user and it increases attack surface.

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+FROM alpine:3.5
+LABEL maintainer Jan Koppe <post@jankoppe.de>
+
+ARG PUMPIO__GUID=1337
```
Not sure that it makes a difference practically but these should probably be system users (<1000), though I do appreciate the choice of 1337 :)

aurbot-user replied:
Yes, the choide of UID/GUID is  really arbitrary in this case. I included it as a build ARG in case the file permissions etc. in deployment differ from the default user. This makes running it in production a bit more comfortable.

Any proposals for a system user UID/GUID?

JanKoppe replied:
Yes, it could be a system user. I don't really see if/how this would make any practical difference, but I don't have any strong opinion about this. Do you have a proposal for a new UID/GUID?

I intentionally made this a build arg, because sometimes it's more comfortable to change the UID/GUID for running in production (file permissions, etc)

strugee replied:
@JanKoppe I wasn't suggesting it not be a build argument, just that the default should be a system user. I don't know if this makes any practical difference either, but it seems like we might as well play it safe :)

Maybe make it 42? Not sure if that's too low (i.e. already occupied); I don't think so though.

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+ARG PUMPIO__GUID=1337
+ARG PUMPIO__UID=1337
+
+ENV PUMP_LOCATION="/usr/local/pumpio"
```
Probably `/usr/local/lib/pumpio` is more standard?

aurbot-user replied:
Yes, you're right!

JanKoppe replied:
Thinking about this a bit more, it should probably be placed in `/usr/lib/node_modules`. Not entirely sure about this though.

strugee replied:
I don't think so. Not if it's not installed directly via `npm install -g`.

strugee commented on docker/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+ARG PUMPIO__UID=1337
+
+ENV PUMP_LOCATION="/usr/local/pumpio"
+ENV PUMP_GIT_TAG="v4.0.0-beta.5"
```
Lop off the beta tag here since 4.0 final shipped since this was sent :)

strugee replied:
Trololololol, I like how @aurbot-user :thumbsup:'d this ;)

JanKoppe commented:
Hi,

thanks a lot for your review strugee! I addressed most of your points in the last commits. One thing to note is that in order to build from the local repository, the Dockerfile et. al. have to live in the root of the repository ( `COPY` statements with parent directories are not allowed).

The last open discussion point as far as I can see is the UID and GUID of the `pumpio` system account.

I would like to leave the UID and GUID configurable via a build argument, this is especially useful when running this in production with a host filesystem location mounted as a volume. This would allow you to easily change the file ownerships to a UID/GUID that suits your environment.

Using IDs < 1000 would be a "cleaner" way, yes. I don't see how the current ID hurts though, so we could just leave it at that. It's a bikeshedding point IMO, but if you have any suggestions for a ID < 1000 I will happily change it! :)

Please excuse the huge amount of stupid spelling mistakes and typos in the original PR, it was quite late when I wrote that (UTC+2) . :)

strugee commented:
@JanKoppe thanks for updating this! Some points (some of which I made inline too, but ¯\_(ツ)_/¯):


> One thing to note is that in order to build from the local repository, the Dockerfile et. al. have to live in the root of the repository ( COPY statements with parent directories are not allowed).

Right. I personally wouldn't mind that. I dunno, maybe this isn't the norm in the Docker space (feel free to tell me if that's the case), but I personally feel like if we ship a Dockerfile in-repo I would expect it to build from what's on the local disk, in the same way that if a project ships e.g. a `Makefile`, I expect that invoking `make` will create a build from local source, not a fresh clone.

> I would like to leave the UID and GUID configurable via a build argument, this is especially useful when running this in production with a host filesystem location mounted as a volume.

Definitely wasn't suggesting it shouldn't be a build argument, just that the default should be different.

> Using IDs < 1000 would be a "cleaner" way, yes. I don't see how the current ID hurts though, so we could just leave it at that. It's a bikeshedding point IMO, but if you have any suggestions for a ID < 1000 I will happily change it! :)

I don't really feel all that strongly about it, but I figured we might as well play it safe :)

Perhaps 42 could serve as an equally amusing UID? :D

strugee commented:
Also wanted to explicitly point out https://github.com/pump-io/pump.io/pull/1348#discussion_r116142318 and https://github.com/pump-io/pump.io/pull/1348#discussion_r116140962 since they're hidden in outdated diffs

JanKoppe commented:
Hey!

took a while to get back on that. I've incorporated your suggestions and prepared the datadir for uploads, etc, which wasn't the case before. 

Sadly `42` was not available, so I just picked a number to move forward.

strugee commented on DOCKER.md:
```diff
@@ -0,0 +1,44 @@
+
+## Running
+
+By default this Dockerfile only includes the `databank-mongodb` driver. Therefore you will have to add the configuration parameters for the MongoDB. The recommended way to configure pump.io is to configure the options via environment variables, as described in the project's README. You can see some example variables in the `docker-compose.yml` file. This file will also show a basic working versio ...
```
"MongoDB" instead of "the MongoDB"

strugee commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+RUN apk add --no-cache graphicsmagick openssl nodejs python make g++ git \
+     && cd "${PUMP_LOCATION}" \
+     && sed -i 's/"bcrypt": "0.8.x"/"bcrypt": "^1.0.2"/g' package.json \
```
This is no longer needed :)

strugee commented:
@JanKoppe ping?

JanKoppe commented:
Dang, this has completely slipped my mind. Will take a look at it tomorrow as it's getting quite late here!

JanKoppe commented:
Hope this is how you want it. If you have any other change request, just force push to my branch, that's fine by me. :)

strugee commented:
@JanKoppe thank you _so much_ for working on this! I'll merge when Travis is green :)
